### PR TITLE
Plotting fix - matched peaks highlighting

### PR DESCRIPTION
**Pull request by matteopilz**
## Description

This should fix the issue of not being able to highlight matching peaks. I have removed the alignment parameter and left matched_peaks as keyword argument. This is to ensure, that the correct matching list for each spectrum is provided, as we have currently no way of enforcing it.

We can keep this PR open, until I have found a way to also make use of the matching from SpectrumAnnotator.

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

**Comment by matteopilz** [comment]
@jpfeuffer, are you sure the matching peaks are stored in the spectrum? Our documentation says, they are placed in PeptideIdentification https://openms.de/current_doxygen/html/classOpenMS_1_1SpectrumAnnotator.html.

**Comment by jpfeuffer** [comment]
I interpret the documentation differently and can confirm that the matches are in the stringdataarray of the spectrum, as expected:
https://openms.de/current_doxygen/html/classOpenMS_1_1SpectrumAnnotator.html#a1acc9d248696c0fd315d476bccbd4e0e

**Comment by matteopilz** [comment]
I see, do you have some example code, so I can look at the way it is stored?

**Comment by jpfeuffer** [comment]
Sure. The annotations are just in the first StringArray of the spectrum, as usual.

```python
import pyopenms as oms
import pyopenms.plotting as msplt
import pyopenms.Constants as msconst
import matplotlib.pyplot as plt

if __name__ == '__main__':
    print(oms.__version__)
    exp = oms.MSExperiment()
    oms.MascotGenericFile().load("xyz.mgf", exp)
    spec = exp[0]
    annot = oms.SpectrumAnnotator()
    tg = oms.TheoreticalSpectrumGenerator()

    tg_params = tg.getDefaults()
    update_params = {
            "add_metainfo": "true",
            "add_a_ions": "true",
            "add_precursor_peaks": "true",
            "add_all_precursor_charges": "true",
            "add_abundant_immonium_ions": "true",
            "add_losses": "false",
            "add_first_prefix_ion": "true",
        }
    for k, v in update_params.items():
        tg_params.setValue(k, v)

    tg.setParameters(tg_params)
    aaseq = oms.AASequence.fromString('(TMT6plex).LTQLGTFEDHFLSLQR')
    theor = oms.MSSpectrum()
    tg.getSpectrum(theor, aaseq, 1, 3)


    peptide = oms.PeptideHit(1.0, 1, 3, aaseq)
    alignment = oms.SpectrumAlignment()
    alignment_params = alignment.getDefaults()
    alignment_params.setValue("tolerance", 100.0)
    alignment_params.setValue("is_relative_tolerance", "true")
    alignment.setParameters(alignment_params)
    annot.annotateMatches(spec, peptide, tg, alignment)
    msplt.plot_spectrum(spec, color_ions=True)
    stop = timeit.default_timer()
    plt.show()
```

**Comment by matteopilz** [comment]
So matching a theoretical spectrum and experimental leaves the experimental spectrum with the matching ions of the theoretical by name and not by index?

**Comment by jpfeuffer** [comment]
The theoretical spectrum is only temporary in this case. I do not need a mirror plot since the predicted intensities are uniform anyway.
It worked in my code. I just thought you found that something else is not working anymore.
If I misunderstood, you can just merge my code. I just wanted to make sure that both use cases will work in the future.

**Comment by timosachsenberg** [comment]
Don’t we also have ids with fragment annotation? Are these also supported ?I think this is very useful @Arslan-Siraj didn‘t you also develop something for NuXL

**Comment by matteopilz** [comment]
Right now we have these options covered for comparing mirror plots:
1. Matching peaks can be highlighted with a matched_peaks set in each spectrum generated from getSpectrumAlignment. This will overwrite color_ions, because it kind of contradicts it. annotate_ions still works fine with it. It's also possible to highlight matching peaks in one spectrum only and ions in the other.
2. We can run SpectrumAnnotator on an experimental spectrum to annotate matches, which will give us colored ions in both spectra, if we want to.

What we don't have right now is a way to only color matching peaks after annotating them with SpectrumAnnotator. Do we want to have that as well?

**Comment by timosachsenberg** [comment]
> What we don't have right now is a way to only color matching peaks after annotating them with SpectrumAnnotator. Do we want to have that as well?

I think the bare minimum we should support:

1. most users just want to plot a spectrum and optionally add some labels to be plotted
2. some users want to visualize identification results:
   - either by using SpectrumAlignment to align a theoretical spectrum of a peptide/metabolite with an experimental one
   - by directly using the https://github.com/OpenMS/OpenMS/blob/develop/src/pyOpenMS/pxds/PeptideHit.pxd#L63 stored in the identification object.
   both with/without mirror plot, supporting labels

There should be an option to:
- set the color for non-matching / non-annotated peaks (e.g., to plot uninteresting peaks with light grey color)
- set the color of matching / annotated peaks (color or "auto"). auto should color b,y ions etc. differently.
- show / hide a sequence ladder (https://github.com/bittremieux/spectrum_utils/issues/24#issuecomment-1620647853)  (e.g. only active if a peptide sequence is manually set - e.g. this allows to set the sequence and annotation from a PeptideHit)

ideally all shown on a tutorial page in pyOpenMS readthedocs

for the future:
- think about metabolite structure annotation

@matteopilz would that be something for a (multiple) sprint task? And @Arslan-Siraj can you provide scripts and input from your work on the ID annotations?

**Comment by jpfeuffer** [comment]
@matteopilz Did you try just disabling mirror plots? What you are saying that doesn't work is exactly why I created my PR.
Having labels from spectrum annotator or somewhere else and just plot ONE spectrum where the corresponding peaks are labelled and colored if they are from an ion series.
This was already working in my PR

**Comment by matteopilz** [comment]
@jpfeuffer, you mean like this? Yes, it works, though I have also based my PR on yours, so some changes are already in there.
image

**Comment by jpfeuffer** [comment]
Yes right!
I was just confused by:

> What we don't have right now is a way to only color matching peaks after annotating them with SpectrumAnnotator. Do we want to have that as well?

**Comment by jpfeuffer** [comment]
Because the plot looks like exactly that.

**Comment by matteopilz** [comment]
You are right, that was a bit unclear, I was referring to mirror plots only. We can toggle color and annotate peaks independently in each spectrum, but we have currently no way of **only** highlighting matching peaks with the info from SpectrumAnnotator. We explicitly need the matching peaks argument.
But Timo is right, there should be an option to color all matching peaks uniformly or automatically, e.g. by ion.

**Comment by Arslan-Siraj** [comment]
I just plot ms2spectra in streamlit app (just from the annotation in idxml file). currently, doing in how we can show more user friendly.

**Comment by jpfeuffer** [comment]
@matteopilz That should be fine. If you activate mirror_plot you should (or IMO must) have a matched_peaks list (unless you do the matching inside the plotting function where you then have to expose all options, which I think does not make sense).

**Comment by matteopilz** [comment]
Highlight matching peaks

```
mirror_plot_spectrum(
    observed_spectrum,
    theo_spectrum,
    spectrum_top_kws={'annotate_ions': False, 'color_ions': False, 'matched_peaks': matched_peaks_observed},
    spectrum_bottom_kws={'annotate_ions': False, 'color_ions': False, 'matched_peaks': matched_peaks_theo}
)
```

image

Change color of peaks

```
mirror_plot_spectrum(
    observed_spectrum,
    theo_spectrum,
    spectrum_top_kws={'annotate_ions': False, 'color_ions': False, 'matched_peaks': matched_peaks_observed, 'colormap': {'matched': 'blue'}},
    spectrum_bottom_kws={'annotate_ions': False, 'color_ions': False, 'matched_peaks': matched_peaks_theo}
)
```

image

We should probably think about how to pass the matched peaks and coloring, right now it's per spectrum.

**Comment by jpfeuffer** [comment]
Looks good. What happens if you have ions annotated (and colored) at a matched peak? Which color does it get?
In general, have you tried annotating ions in a mirror plot yet, just to see if it works (in both plots)?

**Comment by matteopilz** [comment]
Then they are colored based on the ion type, while all unmatched peaks are de-highlighted.
image

We can also change the color of unmatched peaks.
```
mirror_plot_spectrum(
    observed_spectrum,
    theo_spectrum,
    spectrum_top_kws={'annotate_ions': True, 'color_ions': True, 'matched_peaks': matched_peaks_observed, 'colormap': {'unmatched': 'pink'}},
    spectrum_bottom_kws={'annotate_ions': True, 'color_ions': True, 'matched_peaks': matched_peaks_theo, 'colormap': {'unmatched': 'pink'}}
)
```
image

**Comment by jpfeuffer** [comment]
Hmm not bad but do you think it makes sense to allow "by_iontype" as an unmatched "color". Could be useful for the lower spectrum.

**Comment by matteopilz** [comment]
That would be the default when we omit the alignment. There won't be a difference between matched and unmatched. Otherwise we could apply some fixed alpha value to them.

**Comment by Arslan-Siraj** [comment]
def plot_spectra_by_pepthits(h):
    peak_annotation = h.getPeakAnnotations()
    intensity_values_ = []  # To store peak.intensity values
    mz_values_ = []  # To store peak.mz values
    annotations_ = [] # To store ions annotations

    for peak in peak_annotation:
        intensity_values_.append(peak.intensity)
        mz_values_.append(peak.mz)
        annotations_.append(str(peak.annotation))

    spectrum = oms.MSSpectrum()
    spectrum.set_peaks([mz_values_, intensity_values_ ])

    # Sort the peaks according to ascending mass-to-charge ratio
    spectrum.sortByPosition()

    indices = set(range(len(mz_values_)))

    sda = StringDataArray()
    sda.setName("Peak annotation")
    for element in annotations_:
        sda.push_back(element)

    spectrum.setStringDataArrays([sda])
    plot_spectrum(spectrum, annotate_ions=True, matched_peaks=indices) 

import pyopenms as oms
from pyopenms import *
prot_ids = []; pep_ids = []
IdXMLFile().load(str("LWelp_MRaabe_110621_160621_Ecoli_S30_NM_bRP_Frctn1_perc_0.0100_XLs.idXML"), prot_ids, pep_ids)
for peptide_id in pep_ids:
   hits = peptide_id.getHits()
   plot_spectra_by_pepthits(hits[0])
   break

Screenshot 2023-12-04 143544
@matteopilz

**Comment by jpfeuffer** [comment]
> That would be the default when we omit the alignment. There won't be a difference between matched and unmatched. Otherwise we could apply some fixed alpha value to them.

Can you show a short code sample with a picture of what you mean?

**Comment by matteopilz** [comment]
Sure,
that is with the matching peaks, de highlighting all unmatched peaks automatically. 
```
mirror_plot_spectrum(
    observed_spectrum,
    theo_spectrum,
    spectrum_top_kws={'annotate_ions': True, 'color_ions': True, 'matched_peaks': matched_peaks_observed},
    spectrum_bottom_kws={'annotate_ions': True, 'color_ions': True, 'matched_peaks': matched_peaks_theo}
)
```
image

That is without matching peaks, coloring matched and unmatched peaks by ion.
```
mirror_plot_spectrum(
    observed_spectrum,
    theo_spectrum,
    spectrum_top_kws={'annotate_ions': True, 'color_ions': True},
    spectrum_bottom_kws={'annotate_ions': True, 'color_ions': True}
)
```
image

**Comment by jpfeuffer** [comment]
That looks good!

I guess the only thing that people sometimes might want more, is to change color/alpha or even hiding of unmatched peaks. Or in the single spectrum case, the same for the unannotated peaks.

I personally do not need that yet but it might come up.

**Comment by timosachsenberg** [comment]
Using alpha might be less often used for publications. I think a good default is to use light-gray for unmatched peaks and black for single spectrum unannotated.

**Comment by matteopilz** [comment]
Ok, that is currently the default as well.

File: src/pyOpenMS/pyopenms/plotting.py [hunk 1]
```diff
@@ -218,7 +212,7 @@ def plot_chromatogram(c: "MSChromatogram"):
 
 def _annotate_ion(mz: float, intensity: float, annotation: Optional[str],
                   color_ions: bool, annotate_ions: bool, matched: Optional[bool],
-                  annotation_kws: Dict[str, object], ax) -> Tuple[str, int]:
+                  annotation_kws: Dict[str, object], colormap: Dict[str, str], ax) -> Tuple[str, int]:
```
Comment (JeeH-K) [inline src/pyOpenMS/pyopenms/plotting.py:215]: Maybe a short description about the color map is needed in the following comment section.

**Comment by jpfeuffer** [comment]
> Using alpha might be less often used for publications. I think a good default is to use light-gray for unmatched peaks and black for single spectrum unannotated.

The alpha suggestion was mainly to allow alpha=0 to hide unannotated things. You can also add another parameter to allow that, if you want to allow it.

**Review by JeeH-K — APPROVED** [review]

**Pull request merged**